docs/Style.thy

```diff
@@ -219,6 +219,102 @@ lemma test_lemma3:
          case_tac h; simp)
   done
 
+section \<open>Right vs left operator-wrapping\<close>
+
+text \<open>
+  When a term is too long, there is a general consensus to wrap it at operators. However, consensus
+  has never emerged on whether the operator should then end up at the end of the line (right
+  operator wrapping), or start of the next one (left operator wrapping).
+  Mathematicians have a tendency towards right-wrapping, programmers towards left-wrapping, they
+  each have advantages in specific contexts, thus both appear in the l4v proofs and are permitted
+  style.\<close>
+
+term \<open>A \<and> B \<longrightarrow> C \<and> D\<close> \<comment> \<open>no wrapping\<close>
+
+term \<open>A \<and>
+      B \<longrightarrow>
+        C \<and>
+        D\<close> \<comment> \<open>right-wrapping\<close>
+
+term \<open>A
+      \<and> B
+      \<longrightarrow> C
+         \<and> D\<close> \<comment> \<open>left-wrapping\<close>
+
+text \<open>
+  While both styles are permitted, but do not mix them in the same lemma. If a lemma already uses
+  one style and you aren't doing a major rewrite, stick to the existing style.\<close>
+
+lemma
+  shows "\<lbrakk> A; B; C\<rbrakk> \<Longrightarrow>
+         D" \<comment> \<open>right-wrapping OK\<close>
+  and "\<lbrakk> A; B; C\<rbrakk>
+       \<Longrightarrow> D" \<comment> \<open>left-wrapping OK\<close>
+  oops \<comment> \<open>mixing styles: NOT OK\<close>
+
+text \<open>
+  Some operators and syntax only have ONE style. As seen in other sections:
+  * the `|` in `case`  never appears on the right
+  * `;` is always on the right when wrapping lists of assumptions
+  * `shows .. and ... and` wraps with `and` on the left
+  * commas and colons, owing to our natural language origins, always end up on the right\<close>
+
+lemma
+  shows
+    "\<lbrakk> A
+     ; B \<rbrakk> \<comment> \<open>wrong: always on right\<close>
+     \<comment> \<open>ok: can be either\<close>
+     \<Longrightarrow> C" and \<comment> \<open>wrong: `shows/and` only on left!\<close>
+    "D"
+  and "E" \<comment> \<open>ok: on left\<close>
+proof -
+  have "True \<and> True"
+    by (rule conjI,
+        blast,
+        blast) \<comment> \<open>ok\<close>
+  have "True \<and> True"
+    by (rule conjI
+        , blast
+        , blast) \<comment> \<open>NOT OK: commas go on right\<close>
+  have "True \<and> True"
+    by (rule conjI;
+        blast) \<comment> \<open>ok\<close>
+  have "True \<and> True"
+    by (rule conjI
+        ; blast) \<comment> \<open>also ok, can be either, but prefer ; on right if it's a simple method
```

Comment:
I was told that the semicolon must also go at the end of the line

Reply:
I certainly much prefer that, but thought I lost that argument :-). If there is a majority for `;` at the end of line, I'm all for it.

Reply:
I guess it's down to what we want to permit. I examined the situation now that I'm awake again, and it looks like a bit over 100 uses of a leading `;` in `l4v`, so it looks like this one can be arbitrarily permitted or scrapped. What do people think?

In my looking around I also discovered that `|` is rare at end of lines, so maybe we should say that it should go on the left while we're at it?

Then a candidate proof would look something like:
```
apply (a, b,
       aa; bb;
       (c
        | d
        | e
        | f)+)
```
Is this acceptable? Consider variants of all of these methods being short/long. Also, are there any other proof operators I'm forgetting? Obviously `?` and `+` are unary, thus go at the end and don't wrap.

Reply:
It's probably mostly due to just being what I'm used to, but I'm happy with `|` going on the left and `;` being preferred on the right. I know that some people have made the argument that the two are used similarly and should have the same style, but I think I can invent enough of a rationalisation that I am happy with them being different. In particular, to me the first tactic is more important than the second with `;`, while with `|` they're of equivalent importance.

Reply:
I'd be fine with that, yes.